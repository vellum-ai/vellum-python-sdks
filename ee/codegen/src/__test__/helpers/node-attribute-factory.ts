import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
