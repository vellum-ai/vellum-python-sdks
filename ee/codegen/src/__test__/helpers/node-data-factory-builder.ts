import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
