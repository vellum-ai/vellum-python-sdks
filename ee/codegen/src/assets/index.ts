// This directory contains generated assets that are copied during postinstall
// and included in the npm package build. The node-definitions.json file is
// generated from the Python SDK and copied here for use in code generation.
