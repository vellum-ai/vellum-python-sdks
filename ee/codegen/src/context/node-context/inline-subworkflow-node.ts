import { VellumVariableType } from "vellum-ai/api";

import { BaseNodeContext } from "./base";

import { PortContext } from "src/context/port-context";
import { NodeDefinitionGenerationError } from "src/generators/errors";
import { SubworkflowNode as SubworkflowNodeType } from "src/types/vellum";

export class InlineSubworkflowNodeContext extends BaseNodeContext<SubworkflowNodeType> {
  baseNodeClassName = "InlineSubworkflowNode";
  baseNodeDisplayClassName = "BaseInlineSubworkflowNodeDisplay";
  isCore = true;

  getNodeOutputNamesById(): Record<string, string> {
    const subworkflowNodeData = this.nodeData.data;
    if (subworkflowNodeData.variant !== "INLINE") {
      throw new NodeDefinitionGenerationError(
        `SubworkflowNode only supports INLINE variant. Received: ${this.nodeData.data.variant}`
      );
    }

    return subworkflowNodeData.outputVariables.reduce((acc, variable) => {
      acc[variable.id] = variable.key;
      return acc;
    }, {} as Record<string, string>);
  }

  getNodeOutputTypesById(): Record<string, VellumVariableType> {
    const subworkflowNodeData = this.nodeData.data;
    if (subworkflowNodeData.variant !== "INLINE") {
      throw new NodeDefinitionGenerationError(
        `SubworkflowNode only supports INLINE variant. Received: ${this.nodeData.data.variant}`
      );
    }

    return Object.fromEntries(
      subworkflowNodeData.outputVariables.map((variable) => [
        variable.id,
        variable.type,
      ])
    );
  }

  createPortContexts(): PortContext[] {
    return [
      new PortContext({
        workflowContext: this.workflowContext,
        nodeContext: this,
        portId: this.nodeData.data.sourceHandleId,
      }),
    ];
  }
}
