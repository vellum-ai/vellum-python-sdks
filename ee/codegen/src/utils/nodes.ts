import { VellumError } from "vellum-ai/errors";

import { WorkflowNode } from "src/types/vellum";

export function getNodeLabel(nodeData: WorkflowNode): string {
  switch (nodeData.type) {
    case "GENERIC":
      return nodeData.definition?.name ?? nodeData.label ?? "Generic Node";
    default:
      return nodeData.data.label;
  }
}

export function isUnaryOperator(operator: string): boolean {
  return operator === "null" || operator === "notNull";
}

export function isVellumErrorWithDetail(
  error: unknown
): error is VellumError & { body: { detail: string } } {
  return (
    error instanceof VellumError &&
    typeof error.body === "object" &&
    error.body !== null &&
    "detail" in error.body &&
    typeof error.body.detail === "string"
  );
}
