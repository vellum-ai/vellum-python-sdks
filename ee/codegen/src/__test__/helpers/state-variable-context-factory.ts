import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
