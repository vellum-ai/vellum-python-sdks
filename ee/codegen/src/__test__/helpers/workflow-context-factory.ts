import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
