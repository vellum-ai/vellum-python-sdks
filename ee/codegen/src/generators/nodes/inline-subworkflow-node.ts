import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { OUTPUTS_CLASS_NAME } from "src/constants";
import { InlineSubworkflowNodeContext } from "src/context/node-context/inline-subworkflow-node";
import { NodeDefinitionGenerationError } from "src/generators/errors";
import { BaseNestedWorkflowNode } from "src/generators/nodes/bases/nested-workflow-base";
import { WorkflowProjectGenerator } from "src/project";
import {
  SubworkflowNode as SubworkflowNodeType,
  WorkflowRawData,
} from "src/types/vellum";

export class InlineSubworkflowNode extends BaseNestedWorkflowNode<
  SubworkflowNodeType,
  InlineSubworkflowNodeContext
> {
  getInnerWorkflowData(): WorkflowRawData {
    if (this.nodeData.data.variant !== "INLINE") {
      throw new NodeDefinitionGenerationError(
        `InlineSubworkflowNode only supports INLINE variant. Received: ${this.nodeData.data.variant}`
      );
    }

    return this.nodeData.data.workflowRawData;
  }

  getNodeClassBodyStatements(): AstNode[] {
    const nestedWorkflowContext = this.getNestedWorkflowContextByName(
      BaseNestedWorkflowNode.subworkflowNestedProjectName
    );

    const nestedWorkflowReference = python.reference({
      name: nestedWorkflowContext.workflowClassName,
      modulePath: nestedWorkflowContext.modulePath,
    });

    const subworkflowField = python.field({
      name: "subworkflow",
      initializer: nestedWorkflowReference,
    });

    const outputsClass = this.generateOutputsClass();

    return [subworkflowField, outputsClass];
  }

  getNodeDisplayClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    statements.push(
      python.field({
        name: "label",
        initializer: python.TypeInstantiation.str(this.nodeData.data.label),
      })
    );

    statements.push(
      python.field({
        name: "node_id",
        initializer: python.TypeInstantiation.uuid(this.nodeData.id),
      })
    );

    statements.push(
      python.field({
        name: "target_handle_id",
        initializer: python.TypeInstantiation.uuid(
          this.nodeData.data.targetHandleId
        ),
      })
    );

    statements.push(
      python.field({
        name: "workflow_input_ids_by_name",
        initializer: python.TypeInstantiation.dict([]),
      })
    );

    return statements;
  }

  private generateOutputsClass(): python.Class {
    const nodeBaseClassRef = this.getNodeBaseClass();
    const outputsClass = python.class_({
      name: OUTPUTS_CLASS_NAME,
      extends_: [
        python.reference({
          name: nodeBaseClassRef.name,
          modulePath: nodeBaseClassRef.modulePath,
          alias: nodeBaseClassRef.alias,
          attribute: [OUTPUTS_CLASS_NAME],
        }),
      ],
    });

    const nestedWorkflowContext = this.getNestedWorkflowContextByName(
      BaseNestedWorkflowNode.subworkflowNestedProjectName
    );

    nestedWorkflowContext.workflowOutputContexts.forEach((outputContext) => {
      const outputName = outputContext.name;
      const outputField = python.field({
        name: outputName,
        initializer: python.reference({
          name: nestedWorkflowContext.workflowClassName,
          modulePath: nestedWorkflowContext.modulePath,
          attribute: [OUTPUTS_CLASS_NAME, outputName],
        }),
      });

      outputsClass.add(outputField);
    });

    return outputsClass;
  }

  protected getOutputDisplay(): python.Field {
    const nestedWorkflowContext = this.getNestedWorkflowContextByName(
      BaseNestedWorkflowNode.subworkflowNestedProjectName
    );

    return python.field({
      name: "output_display",
      initializer: python.TypeInstantiation.dict(
        nestedWorkflowContext.workflowOutputContexts.map((outputContext) => {
          const outputName = outputContext.name;
          const finalOutput = outputContext.getFinalOutputNodeData();
          let outputId;
          if ("type" in finalOutput) {
            outputId = finalOutput.data.outputId;
          } else {
            outputId = finalOutput.value.nodeOutputId;
          }

          return {
            key: python.reference({
              name: this.nodeContext.nodeClassName,
              modulePath: this.nodeContext.nodeModulePath,
              attribute: [OUTPUTS_CLASS_NAME, outputName],
            }),
            value: python.instantiateClass({
              classReference: python.reference({
                name: "NodeOutputDisplay",
                modulePath:
                  this.workflowContext.sdkModulePathNames
                    .NODE_DISPLAY_TYPES_MODULE_PATH,
              }),
              arguments_: [
                python.methodArgument({
                  name: "id",
                  value: python.TypeInstantiation.uuid(outputId),
                }),
                python.methodArgument({
                  name: "name",
                  value: python.TypeInstantiation.str(outputName),
                }),
              ],
            }),
          };
        })
      ),
    });
  }

  protected getNestedWorkflowProject(): WorkflowProjectGenerator {
    if (this.nodeData.data.variant !== "INLINE") {
      throw new NodeDefinitionGenerationError(
        `SubworkflowNode only supports INLINE variant. Received: ${this.nodeData.data.variant}`
      );
    }

    const inlineSubworkflowNodeData = this.nodeData.data;
    const nestedWorkflowContext = this.getNestedWorkflowContextByName(
      BaseNestedWorkflowNode.subworkflowNestedProjectName
    );

    return new WorkflowProjectGenerator({
      workflowVersionExecConfig: {
        workflowRawData: inlineSubworkflowNodeData.workflowRawData,
        inputVariables: inlineSubworkflowNodeData.inputVariables,
        outputVariables: inlineSubworkflowNodeData.outputVariables,
      },
      moduleName: nestedWorkflowContext.moduleName,
      workflowContext: nestedWorkflowContext,
    });
  }

  protected getErrorOutputId(): string | undefined {
    return this.nodeData.data.errorOutputId;
  }
}
