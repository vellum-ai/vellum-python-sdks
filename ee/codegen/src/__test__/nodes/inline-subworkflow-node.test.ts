import { mkdir, readFile, rm } from "fs/promises";
import { join } from "path";

import { v4 as uuidv4 } from "uuid";
import { beforeEach } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import {
  inlineSubworkflowNodeDataFactory,
  templatingNodeFactory,
} from "src/__test__/helpers/node-data-factories";
import { makeTempDir } from "src/__test__/helpers/temp-dir";
import { createNodeContext } from "src/context";
import { InlineSubworkflowNodeContext } from "src/context/node-context/inline-subworkflow-node";
import { InlineSubworkflowNode } from "src/generators/nodes/inline-subworkflow-node";

describe("InlineSubworkflowNode", () => {
  let tempDir: string;

  beforeEach(async () => {
    tempDir = makeTempDir("inline-subworkflow-node-test");
    await mkdir(tempDir, { recursive: true });
  });

  afterEach(async () => {
    await rm(tempDir, { recursive: true, force: true });
  });

  describe("basic", () => {
    beforeEach(async () => {
      const workflowContext = workflowContextFactory({
        absolutePathToOutputDirectory: tempDir,
        moduleName: "code",
      });
      const nodeData = inlineSubworkflowNodeDataFactory();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as InlineSubworkflowNodeContext;

      const node = new InlineSubworkflowNode({
        workflowContext,
        nodeContext,
      });

      await node.persist();
    });

    it(`inline subworkflow node file`, async () => {
      expect(
        await readFile(
          join(
            tempDir,
            "code",
            "nodes",
            "inline_subworkflow_node",
            "__init__.py"
          ),
          "utf-8"
        )
      ).toMatchSnapshot();
    });

    it(`inline subworkflow node display file`, async () => {
      expect(
        await readFile(
          join(
            tempDir,
            "code",
            "display",
            "nodes",
            "inline_subworkflow_node",
            "__init__.py"
          ),
          "utf-8"
        )
      ).toMatchSnapshot();
    });
  });

  describe("name collision", () => {
    beforeEach(async () => {
      const workflowContext = workflowContextFactory({
        absolutePathToOutputDirectory: tempDir,
        moduleName: "code",
      });
      const nodeData = inlineSubworkflowNodeDataFactory({
        label: "My node",
        nodes: [templatingNodeFactory({ label: "My node" })],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as InlineSubworkflowNodeContext;

      const node = new InlineSubworkflowNode({
        workflowContext,
        nodeContext,
      });

      await node.persist();
    });

    it(`should handle subworkflow with same name as internal node`, async () => {
      expect(
        await readFile(
          join(tempDir, "code", "nodes", "my_node", "__init__.py"),
          "utf-8"
        )
      ).toMatchSnapshot();

      expect(
        await readFile(
          join(tempDir, "code", "display", "nodes", "my_node", "__init__.py"),
          "utf-8"
        )
      ).toMatchSnapshot();

      expect(
        await readFile(
          join(tempDir, "code", "nodes", "my_node", "nodes", "my_node.py"),
          "utf-8"
        )
      ).toMatchSnapshot();

      expect(
        await readFile(
          join(
            tempDir,
            "code",
            "display",
            "nodes",
            "my_node",
            "nodes",
            "my_node.py"
          ),
          "utf-8"
        )
      ).toMatchSnapshot();
    });
  });
  describe("adornments", () => {
    beforeEach(async () => {
      const workflowContext = workflowContextFactory({
        absolutePathToOutputDirectory: tempDir,
        moduleName: "code",
      });
      const nodeData = inlineSubworkflowNodeDataFactory({
        label: "My node",
        nodes: [templatingNodeFactory({ label: "My node" })],
        adornments: [
          {
            id: uuidv4(),
            label: "RetryNodeLabel",
            base: {
              name: "RetryNode",
              module: [
                "vellum",
                "workflows",
                "nodes",
                "core",
                "retry_node",
                "node",
              ],
            },
            attributes: [
              {
                id: uuidv4(),
                name: "max_attempts",
                value: {
                  type: "CONSTANT_VALUE",
                  value: {
                    type: "NUMBER",
                    value: 3,
                  },
                },
              },
              {
                id: uuidv4(),
                name: "delay",
                value: {
                  type: "CONSTANT_VALUE",
                  value: {
                    type: "NUMBER",
                    value: 2,
                  },
                },
              },
            ],
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as InlineSubworkflowNodeContext;

      const node = new InlineSubworkflowNode({
        workflowContext,
        nodeContext,
      });

      await node.persist();
    });

    it(`should generate adornments`, async () => {
      expect(
        await readFile(
          join(tempDir, "code", "nodes", "my_node", "__init__.py"),
          "utf-8"
        )
      ).toMatchSnapshot();
      expect(
        await readFile(
          join(tempDir, "code", "display", "nodes", "my_node", "__init__.py"),
          "utf-8"
        )
      ).toMatchSnapshot();
    });
  });
});
