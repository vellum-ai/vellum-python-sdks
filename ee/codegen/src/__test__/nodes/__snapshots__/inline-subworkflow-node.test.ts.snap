// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InlineSubworkflowNode > adornments > should generate adornments 1`] = `
"from vellum.workflows.nodes.core.retry_node.node import RetryNode
from vellum.workflows.nodes.displayable import InlineSubworkflowNode

from .workflow import MyNodeWorkflow


@RetryNode.wrap(max_attempts=3, delay=2)
class MyNode(InlineSubworkflowNode):
    subworkflow = MyNodeWorkflow
"
`;

exports[`InlineSubworkflowNode > adornments > should generate adornments 2`] = `
"# flake8: noqa: F401, F403

from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseInlineSubworkflowNodeDisplay, BaseRetryNodeDisplay
from vellum_ee.workflows.display.nodes.types import NodeOutputDisplay, PortDisplayOverrides
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ....nodes.my_node import MyNode
from .nodes import *
from .workflow import *


@BaseRetryNodeDisplay.wrap(node_id=UUID("ae49ef72-6ad7-441a-a20d-76c71ad851ef"))
class MyNodeDisplay(BaseInlineSubworkflowNodeDisplay[MyNode]):
    label = "My node"
    node_id = UUID("14fee4a0-ad25-402f-b942-104d3a5a0824")
    target_handle_id = UUID("3fe4b4a6-5ed2-4307-ac1c-02389337c4f2")
    workflow_input_ids_by_name = {}
    node_input_ids_by_name = {}
    output_display = {
        MyNode.Outputs.final_output: NodeOutputDisplay(
            id=UUID("edd5cfd5-6ad8-437d-8775-4b9aeb62a5fb"), name="final-output"
        )
    }
    port_displays = {MyNode.Ports.default: PortDisplayOverrides(id=UUID("4878f525-d4a3-4e3d-9221-e146f282a96a"))}
    display_data = NodeDisplayData(position=NodeDisplayPosition(x=0, y=0), width=None, height=None)
"
`;

exports[`InlineSubworkflowNode > basic > inline subworkflow node display file 1`] = `
"# flake8: noqa: F401, F403

from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseInlineSubworkflowNodeDisplay
from vellum_ee.workflows.display.nodes.types import NodeOutputDisplay, PortDisplayOverrides
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ....nodes.inline_subworkflow_node import InlineSubworkflowNode
from .nodes import *
from .workflow import *


class InlineSubworkflowNodeDisplay(BaseInlineSubworkflowNodeDisplay[InlineSubworkflowNode]):
    label = "Inline Subworkflow Node"
    node_id = UUID("14fee4a0-ad25-402f-b942-104d3a5a0824")
    target_handle_id = UUID("3fe4b4a6-5ed2-4307-ac1c-02389337c4f2")
    workflow_input_ids_by_name = {}
    node_input_ids_by_name = {}
    output_display = {
        InlineSubworkflowNode.Outputs.final_output: NodeOutputDisplay(
            id=UUID("edd5cfd5-6ad8-437d-8775-4b9aeb62a5fb"), name="final-output"
        )
    }
    port_displays = {
        InlineSubworkflowNode.Ports.default: PortDisplayOverrides(id=UUID("4878f525-d4a3-4e3d-9221-e146f282a96a"))
    }
    display_data = NodeDisplayData(position=NodeDisplayPosition(x=0, y=0), width=None, height=None)
"
`;

exports[`InlineSubworkflowNode > basic > inline subworkflow node file 1`] = `
"from vellum.workflows.nodes.displayable import InlineSubworkflowNode as BaseInlineSubworkflowNode

from .workflow import InlineSubworkflowNodeWorkflow


class InlineSubworkflowNode(BaseInlineSubworkflowNode):
    subworkflow = InlineSubworkflowNodeWorkflow
"
`;

exports[`InlineSubworkflowNode > name collision > should handle subworkflow with same name as internal node 1`] = `
"from vellum.workflows.nodes.displayable import InlineSubworkflowNode

from .workflow import MyNodeWorkflow


class MyNode(InlineSubworkflowNode):
    subworkflow = MyNodeWorkflow
"
`;

exports[`InlineSubworkflowNode > name collision > should handle subworkflow with same name as internal node 2`] = `
"# flake8: noqa: F401, F403

from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseInlineSubworkflowNodeDisplay
from vellum_ee.workflows.display.nodes.types import NodeOutputDisplay, PortDisplayOverrides
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ....nodes.my_node import MyNode
from .nodes import *
from .workflow import *


class MyNodeDisplay(BaseInlineSubworkflowNodeDisplay[MyNode]):
    label = "My node"
    node_id = UUID("14fee4a0-ad25-402f-b942-104d3a5a0824")
    target_handle_id = UUID("3fe4b4a6-5ed2-4307-ac1c-02389337c4f2")
    workflow_input_ids_by_name = {}
    node_input_ids_by_name = {}
    output_display = {
        MyNode.Outputs.final_output: NodeOutputDisplay(
            id=UUID("edd5cfd5-6ad8-437d-8775-4b9aeb62a5fb"), name="final-output"
        )
    }
    port_displays = {MyNode.Ports.default: PortDisplayOverrides(id=UUID("4878f525-d4a3-4e3d-9221-e146f282a96a"))}
    display_data = NodeDisplayData(position=NodeDisplayPosition(x=0, y=0), width=None, height=None)
"
`;

exports[`InlineSubworkflowNode > name collision > should handle subworkflow with same name as internal node 3`] = `
"from vellum.workflows.nodes.displayable import TemplatingNode
from vellum.workflows.state import BaseState


class MyNode1(TemplatingNode[BaseState, str]):
    template = """Hello, World!"""
    inputs = {}
"
`;

exports[`InlineSubworkflowNode > name collision > should handle subworkflow with same name as internal node 4`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseTemplatingNodeDisplay
from vellum_ee.workflows.display.nodes.types import NodeOutputDisplay, PortDisplayOverrides
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from .....nodes.my_node.nodes.my_node import MyNode1


class MyNode1Display(BaseTemplatingNodeDisplay[MyNode1]):
    label = "My node"
    node_id = UUID("46e221ab-a749-41a2-9242-b1f5bf31f3a5")
    target_handle_id = UUID("3960c8e1-9baa-4b9c-991d-e399d16a45aa")
    template_input_id = UUID("7b8af68b-cf60-4fca-9c57-868042b5b616")
    node_input_ids_by_name = {"template": UUID("7b8af68b-cf60-4fca-9c57-868042b5b616")}
    output_display = {
        MyNode1.Outputs.result: NodeOutputDisplay(id=UUID("2d4f1826-de75-499a-8f84-0a690c8136ad"), name="result")
    }
    port_displays = {MyNode1.Ports.default: PortDisplayOverrides(id=UUID("6ee2c814-d0a5-4ec9-83b6-45156e2f22c4"))}
    display_data = NodeDisplayData(position=NodeDisplayPosition(x=0, y=0), width=None, height=None)
"
`;
